### PR TITLE
Review

### PR DESCRIPTION
**Pull request by mllg**

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+License: LGPL-3
+URL: https://mlr3spatiotempcv.mlr-org.com/,
+    https://github.com/mlr-org/mlr3spatiotempcv,
+    https://mlr3book.mlr-org.com
+BugReports: https://github.com/mlr-org/mlr3spatiotempcv/issues
+Depends:
+    R (>= 3.5.0)
+Imports:
+    checkmate,
+    cli,
```
Comment (mllg) [inline DESCRIPTION:31]: I don't think we need cli, we have R's conditions and additionally a logger to output messages.

File: DESCRIPTION [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    paradox,
+    R6,
+    testthat,
+    utils
+Suggests:
+    bbotk,
+    bibtex,
+    blockCV (>= 2.1.1),
+    bookdown,
+    caret,
```
Comment (mllg) [inline DESCRIPTION:45]: I don't think this dependency is justified just for `createFolds`. Is `split(mlr3mis::shuffle(y), 1:k)` sufficient?

File: DESCRIPTION [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    mlr3misc (>= 0.1.7),
+    paradox,
+    R6,
+    testthat,
+    utils
+Suggests:
+    bbotk,
+    bibtex,
+    blockCV (>= 2.1.1),
+    bookdown,
```
Comment (mllg) [inline DESCRIPTION:44]: Don't know why this is in suggests, not needed in the package.

File: NAMESPACE [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import(data.table)
+import(data.table, except = transpose)
+import(ggplot2)
+import(mlr3)
+import(mlr3misc)
+import(paradox)
+importFrom(R6,R6Class)
+importFrom(data.table,rbindlist)
+importFrom(graphics,plot)
+importFrom(mlr3,as_data_backend)
```
Comment (mllg) [inline NAMESPACE:53]: mlr3 already imported

File: R/helper_autoplot.R [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    n_iters = resampling$iters
+  }
+
+  for (i in seq_len(n_iters)) {
+
+    row_id = resampling$instance$test[[i]]
+
+    data$indicator[data$row_id %in% row_id] = i
+  }
+  rm(resampling)
```
Comment (mllg) [inline R/helper_autoplot.R:22]: This is usually not required.

File: DESCRIPTION (hunk 2 above)
Comment (pat-s) [reply DESCRIPTION:45]: Would be great if we can get rid of {caret}, will have a look.

File: DESCRIPTION (hunk 3 above)
Comment (pat-s) [reply DESCRIPTION:44]: Good question, I also don't know. Maybe a c/p leftover.

File: R/helper_autoplot.R (hunk 5 above)
Comment (pat-s) [reply R/helper_autoplot.R:22]: Good catch, no idea why it's there. Maybe a debugging leftover.

File: DESCRIPTION (hunk 1 above)
Comment (pat-s) [reply DESCRIPTION:31]: I like {cli} but yeah I know we do not use it. Will change it.

**Comment by mllg** [comment]
Just to be sure: do not merge the PR! I've only created it for the review. ;)

**Comment by pat-s** [comment]
Not sure why I pushed here, must have been tired 😄

**Comment by pat-s** [comment]
I am confused, I did not push this branch. Is there some auto-update happening? 🤷

File: DESCRIPTION (hunk 2 above)
Comment (pat-s) [reply DESCRIPTION:52]: I think the approach is the same but I cannot reach full reproducibility. It seems caret is doing some internal stochastic things that cannot be reproduced outside?

``` r
library(mlr3misc)
library(caret)
#> Loading required package: lattice
#> Loading required package: ggplot2
library(data.table)

input <- structure(c(
  9L, 7L, 79L, 75L, 231L, 201L, 95L, 19L, 245L, 349L,
  307L, 339L, 3L, 217L, 231L, 61L, 209L, 3L, 245L, 19L, 331L, 61L,
  139L, 75L, 125L, 67L, 205L, 231L, 231L, 217L, 173L, 19L, 61L,
  75L, 339L, 215L, 9L, 173L, 245L, 205L, 361L, 33L, 197L, 201L,
  19L, 19L, 339L, 305L, 67L, 205L, 133L, 173L, 9L, 19L, 61L, 9L,
  79L, 331L, 75L, 215L, 61L, 275L, 237L, 141L, 31L, 125L, 307L,
  33L, 215L, 35L, 119L, 231L, 333L, 237L, 31L, 61L, 305L, 197L,
  173L, 309L, 61L, 135L, 309L, 67L, 9L, 61L, 205L, 201L, 95L, 215L,
  339L, 35L, 75L, 67L, 305L, 231L, 31L, 349L, 133L, 33L, 307L,
  61L, 35L, 231L, 9L, 245L, 307L, 275L, 79L, 209L, 133L, 67L, 67L,
  33L, 33L, 339L, 135L, 349L, 125L, 305L, 231L, 135L, 125L, 309L,
  129L, 19L, 9L, 275L, 333L, 245L, 305L, 33L, 141L, 67L, 31L, 95L,
  135L, 331L, 215L, 231L, 129L, 9L, 31L, 339L, 311L, 129L, 35L,
  125L, 313L, 35L, 245L, 67L, 217L, 307L, 31L, 33L, 245L, 79L,
  61L, 67L, 313L, 209L, 141L, 215L, 95L, 217L, 339L, 173L, 141L,
  35L, 333L, 339L, 119L, 205L, 309L, 245L, 197L, 141L, 163L, 333L,
  119L, 125L, 231L, 35L, 339L, 275L, 133L, 119L, 133L, 197L, 331L,
  31L, 139L, 133L, 139L, 35L, 3L, 309L, 333L, 7L, 311L, 61L, 361L,
  129L, 129L, 231L, 19L, 35L, 197L, 125L, 9L, 79L, 245L, 215L,
  133L, 75L, 33L, 141L, 205L, 333L, 331L, 67L, 333L, 79L, 201L,
  135L, 119L, 135L, 205L, 305L, 95L, 139L, 119L, 163L, 133L, 61L,
  141L, 7L, 125L, 163L, 237L, 3L, 119L, 133L, 33L, 231L, 95L, 311L,
  201L, 3L, 333L, 61L, 231L, 31L, 231L, 9L, 9L, 331L, 331L, 349L,
  311L, 349L, 331L, 19L, 129L, 205L, 129L, 119L, 339L, 61L, 197L,
  231L, 79L, 173L, 245L, 95L, 163L, 205L, 7L, 95L, 3L, 79L, 75L,
  135L, 95L, 163L, 141L, 349L, 33L, 133L, 139L, 313L, 163L, 215L,
  231L, 361L, 217L, 125L, 129L, 173L, 7L, 197L, 129L, 75L, 275L,
  237L, 173L, 215L, 311L, 33L, 205L, 67L, 331L, 19L, 173L, 349L,
  79L, 141L, 173L, 75L, 133L, 61L, 307L, 309L, 209L, 197L, 331L,
  237L, 3L, 7L, 9L, 133L, 205L, 67L, 135L, 135L, 201L, 349L, 349L,
  311L, 67L, 61L, 31L, 95L, 309L, 361L, 309L, 95L, 75L, 173L, 75L,
  215L, 237L, 163L, 275L, 19L, 125L, 331L, 205L, 139L, 313L, 75L,
  361L, 163L, 79L, 19L, 125L, 67L, 135L, 141L, 7L, 215L, 275L,
  245L, 215L, 9L, 19L, 9L, 205L, 333L, 67L, 209L, 307L, 7L, 133L,
  7L, 79L, 197L, 311L, 197L, 141L, 7L, 307L, 217L, 205L, 339L,
  67L, 95L, 201L, 61L, 61L, 339L, 163L, 245L, 311L, 31L, 275L,
  305L, 313L, 197L, 331L, 19L, 339L, 237L, 215L, 205L, 313L, 7L,
  215L, 311L, 339L, 33L, 67L, 349L, 305L, 141L, 163L, 135L, 333L,
  125L, 125L, 75L, 309L, 75L, 119L, 67L, 173L, 139L, 339L, 95L,
  205L, 237L, 33L, 31L, 205L, 209L, 307L, 19L, 61L, 275L, 173L,
  19L, 31L, 119L, 217L, 33L, 79L, 307L, 61L, 3L, 305L, 163L, 35L,
  135L, 135L, 313L, 133L, 61L, 217L, 67L, 139L, 19L, 173L, 61L,
  141L, 163L, 79L, 75L, 119L, 119L, 33L, 7L, 7L, 119L, 173L, 139L,
  237L, 349L, 209L, 275L, 205L, 19L, 119L, 245L, 141L, 349L, 135L,
  35L, 305L, 231L
), .Label = c(
  "CAF001", "CAF002", "CAF003", "CAF004",
  "CAF005", "CAF006", "CAF007", "CAF008", "CAF009", "CAF010", "CAF011",
  "CAF012", "CAF013", "CAF014", "CAF015", "CAF016", "CAF017", "CAF018",
  "CAF019", "CAF020", "CAF021", "CAF022", "CAF023", "CAF024", "CAF025",
  "CAF026", "CAF027", "CAF028", "CAF029", "CAF030", "CAF031", "CAF032",
  "CAF033", "CAF034", "CAF035", "CAF036", "CAF037", "CAF038", "CAF039",
  "CAF040", "CAF041", "CAF042", "CAF043", "CAF044", "CAF045", "CAF046",
  "CAF047", "CAF048", "CAF049", "CAF050", "CAF051", "CAF052", "CAF053",
  "CAF054", "CAF055", "CAF056", "CAF057", "CAF058", "CAF059", "CAF060",
  "CAF061", "CAF062", "CAF063", "CAF064", "CAF065", "CAF066", "CAF067",
  "CAF068", "CAF069", "CAF070", "CAF071", "CAF072", "CAF073", "CAF074",
  "CAF075", "CAF076", "CAF077", "CAF078", "CAF079", "CAF080", "CAF081",
  "CAF082", "CAF083", "CAF084", "CAF085", "CAF086", "CAF087", "CAF088",
  "CAF089", "CAF090", "CAF091", "CAF092", "CAF093", "CAF094", "CAF095",
  "CAF096", "CAF097", "CAF098", "CAF099", "CAF100", "CAF101", "CAF102",
  "CAF103", "CAF104", "CAF105", "CAF106", "CAF107", "CAF108", "CAF109",
  "CAF110", "CAF111", "CAF112", "CAF113", "CAF114", "CAF115", "CAF116",
  "CAF117", "CAF118", "CAF119", "CAF120", "CAF121", "CAF122", "CAF123",
  "CAF124", "CAF125", "CAF126", "CAF127", "CAF128", "CAF129", "CAF130",
  "CAF131", "CAF132", "CAF133", "CAF134", "CAF135", "CAF136", "CAF137",
  "CAF138", "CAF139", "CAF140", "CAF141", "CAF142", "CAF143", "CAF144",
  "CAF145", "CAF146", "CAF147", "CAF148", "CAF149", "CAF150", "CAF151",
  "CAF152", "CAF153", "CAF154", "CAF155", "CAF156", "CAF157", "CAF158",
  "CAF159", "CAF160", "CAF161", "CAF162", "CAF163", "CAF164", "CAF165",
  "CAF166", "CAF167", "CAF168", "CAF169", "CAF170", "CAF171", "CAF172",
  "CAF173", "CAF174", "CAF175", "CAF176", "CAF177", "CAF178", "CAF179",
  "CAF180", "CAF181", "CAF182", "CAF183", "CAF184", "CAF185", "CAF186",
  "CAF187", "CAF188", "CAF189", "CAF190", "CAF191", "CAF192", "CAF193",
  "CAF194", "CAF195", "CAF196", "CAF197", "CAF198", "CAF199", "CAF200",
  "CAF201", "CAF202", "CAF203", "CAF204", "CAF205", "CAF206", "CAF207",
  "CAF208", "CAF209", "CAF210", "CAF211", "CAF212", "CAF213", "CAF214",
  "CAF215", "CAF216", "CAF217", "CAF218", "CAF219", "CAF220", "CAF221",
  "CAF222", "CAF223", "CAF224", "CAF225", "CAF226", "CAF227", "CAF228",
  "CAF229", "CAF230", "CAF231", "CAF232", "CAF233", "CAF234", "CAF235",
  "CAF236", "CAF237", "CAF238", "CAF239", "CAF240", "CAF241", "CAF242",
  "CAF243", "CAF244", "CAF245", "CAF246", "CAF247", "CAF248", "CAF249",
  "CAF250", "CAF251", "CAF252", "CAF253", "CAF254", "CAF255", "CAF256",
  "CAF257", "CAF258", "CAF259", "CAF260", "CAF261", "CAF262", "CAF263",
  "CAF264", "CAF265", "CAF266", "CAF267", "CAF268", "CAF269", "CAF270",
  "CAF271", "CAF272", "CAF273", "CAF274", "CAF275", "CAF276", "CAF277",
  "CAF278", "CAF279", "CAF280", "CAF281", "CAF282", "CAF283", "CAF284",
  "CAF285", "CAF286", "CAF287", "CAF288", "CAF289", "CAF290", "CAF291",
  "CAF292", "CAF293", "CAF297", "CAF298", "CAF299", "CAF300", "CAF301",
  "CAF302", "CAF303", "CAF304", "CAF305", "CAF306", "CAF307", "CAF308",
  "CAF309", "CAF310", "CAF311", "CAF312", "CAF313", "CAF314", "CAF315",
  "CAF316", "CAF323", "CAF324", "CAF325", "CAF326", "CAF327", "CAF328",
  "CAF329", "CAF330", "CAF331", "CAF332", "CAF333", "CAF334", "CAF335",
  "CAF336", "CAF337", "CAF338", "CAF348", "CAF349", "CAF350", "CAF351",
  "CAF352", "CAF353", "CAF354", "CAF355", "CAF356", "CAF357", "CAF358",
  "CAF359", "CAF360", "CAF371", "CAF372", "CAF373", "CAF374", "CAF375",
  "CAF376", "CAF377", "CAF378", "CAF379", "CAF380", "CAF381", "CAF394",
  "CAF395", "CAF396", "CAF397", "CAF398", "CAF399", "CAF400", "CAF401",
  "CAF402", "CAF419", "CAF420", "CAF421", "CAF422", "CAF423", "CAF424",
  "CAF425"
), class = "factor")
k <- 5

set.seed(42)
foo <- mlr3misc::map(
  split(mlr3misc::shuffle(
    1:uniqueN(input),
    uniqueN(input)
  ), 1:k), function(y) {
    unique(input)[y]
  }
)
#> Warning in split.default(mlr3misc::shuffle(1:uniqueN(input), uniqueN(input)), :
#> data length is not a multiple of split variable
foo <- mlr3misc::set_names(foo, sprintf("Fold%s", seq(1:k)))
foo
#> $Fold1
#> [1] CAF135 CAF139 CAF217 CAF079 CAF215 CAF129 CAF201 CAF357 CAF349
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold2
#> [1] CAF009 CAF401 CAF067 CAF245 CAF237 CAF316 CAF141 CAF125
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold3
#> [1] CAF033 CAF095 CAF197 CAF075 CAF007 CAF209 CAF173 CAF308
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold4
#> [1] CAF377 CAF314 CAF163 CAF031 CAF019 CAF133 CAF351 CAF119
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold5
#> [1] CAF312 CAF231 CAF061 CAF003 CAF275 CAF310 CAF035 CAF205
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425

set.seed(42)
# care::createFolds() does things a bit different, hence we loose full repr
# the concept should be the same though?
caret <- lapply(caret::createFolds(
  1:uniqueN(input),
  k
), function(y) {
  unique(input)[y]
})
caret
#> $Fold1
#> [1] CAF007 CAF075 CAF357 CAF139 CAF033 CAF275 CAF135 CAF129
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold2
#> [1] CAF095 CAF377 CAF209 CAF067 CAF215 CAF133 CAF314 CAF163
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold3
#> [1] CAF079 CAF231 CAF310 CAF061 CAF205 CAF173 CAF141 CAF119 CAF316
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold4
#> [1] CAF201 CAF019 CAF349 CAF125 CAF401 CAF237 CAF031 CAF351
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425
#> 
#> $Fold5
#> [1] CAF009 CAF245 CAF003 CAF217 CAF197 CAF308 CAF035 CAF312
#> 369 Levels: CAF001 CAF002 CAF003 CAF004 CAF005 CAF006 CAF007 CAF008 ... CAF425

all.equal(foo, caret)
#> [1] "Component \"Fold1\": Lengths: 9, 8"                                    
#> [2] "Component \"Fold1\": Lengths (9, 8) differ (string compare on first 8)"
#> [3] "Component \"Fold1\": 8 string mismatches"                              
#> [4] "Component \"Fold2\": 8 string mismatches"                              
#> [5] "Component \"Fold3\": Lengths: 8, 9"                                    
#> [6] "Component \"Fold3\": Lengths (8, 9) differ (string compare on first 8)"
#> [7] "Component \"Fold3\": 8 string mismatches"                              
#> [8] "Component \"Fold4\": 8 string mismatches"                              
#> [9] "Component \"Fold5\": 7 string mismatches"
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Comment (pat-s) [reply DESCRIPTION:51]: @mllg I really cannot find a way to reproduce what `caret::createFolds()` returns.

The goal is to mimic the underlying functionality of {CAST} and even if the sampling approach is correct on our side via `shuffle()`, if users are not able to get the exact same results as the underlying package, this is bad.

Maybe you have time to take a detailed look at it? The code is in https://github.com/mlr-org/mlr3spatiotempcv/blob/master/R/helper_sample.R. You can use the tests [here](https://github.com/mlr-org/mlr3spatiotempcv/blob/master/R/helper_sample.R) for debugging.

**Comment by pat-s** [comment]
@mllg not sure why and how but someone and automatic merge with master is happening here

**Comment by mllg** [comment]
Don't know, but I guess you can just close this PR and delete the branch at some point?

**Comment by pat-s** [comment]
The reason for the updates is that this PR goes from `master` to `review` and hence updates here are mirrored from master.

**Comment by pat-s** [comment]
@mllg All comments addressed. Thanks again for your input! Let's release this.

**Pull request closed without merging**